src/core/eme/index.ts

```diff
@@ -135,53 +91,60 @@ function handleEncryptedEvent(
  * The EME handler can be given one or multiple systems and will choose the
  * appropriate one supported by the user's browser.
  * @param {HTMLMediaElement} video
- * @param {Object} keySystems
+ * @param {Array.<Object>} keySystems
  * @param {Subject} errorStream
  * @returns {Observable}
  */
 function createEME(
   video : HTMLMediaElement,
-  keySystems: IKeySystemOption[],
-  errorStream: ErrorStream
-) : Observable<IMockMediaKeys|MediaKeys|ISessionEvent|Event> {
+  keySystemOptions: IKeySystemOption[],
+  errorStream: Subject<Error|CustomError>
+) : Observable<ISessionEvent> {
   if (__DEV__) {
-    keySystems.forEach((ks) => assertInterface(ks, {
+    keySystemOptions.forEach((option) => assertInterface(option, {
       getLicense: "function",
       type: "string",
     }, "keySystem"));
   }
 
-  // get the MediaKeys element the associated options
-  // This is done as the "encrypted" event is not yet received for performance
-  const mediaKeysInfos$ = findCompatibleKeySystem(keySystems, instanceInfos)
-    .mergeMap((keySystemAccessInfos) => {
-      return createMediaKeysObs(keySystemAccessInfos.keySystemAccess)
-        .mergeMap(function prepareMediaKeysConfiguration(mediaKeys) {
-          const { keySystem } = keySystemAccessInfos;
-          const { serverCertificate } = keySystem;
-          setSessionStorage(keySystem);
-
-          const _mediaKeysInfos$ = Observable
-            .of(objectAssign({ mediaKeys }, keySystemAccessInfos));
-
-          if (
-            serverCertificate != null &&
-            typeof mediaKeys.setServerCertificate === "function"
-          ) {
-            return trySettingServerCertificate(mediaKeys, serverCertificate, errorStream)
-              .concat(_mediaKeysInfos$);
-          }
-          return _mediaKeysInfos$;
-        });
-    });
-
   return Observable.combineLatest(
-    onEncrypted$(video).take(1),
-    mediaKeysInfos$
+    onEncrypted$(video),
+    findCompatibleKeySystem(keySystemOptions, currentMediaKeysInfos)
+      .mergeMap((keySystemInfos) => createMediaKeys(keySystemInfos, errorStream))
   )
-    .mergeMap(([encryptedEvent, mediaKeysInfos]) =>
-      handleEncryptedEvent(encryptedEvent, mediaKeysInfos, video, errorStream)
-    );
+    .concatMap(([encryptedEvent, mediaKeysInfos], i) =>
+      Observable.merge(
+        createSessionForEncryptedEvent(encryptedEvent, mediaKeysInfos),
+        i === 0 ?
+          attachMediaKeys(mediaKeysInfos, video, currentMediaKeysInfos).ignoreElements() :
+          Observable.empty() as Observable<never>
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Remove this unnecessary cast. ([typescript:S4325](https://sonarcloud.io/organizations/rx-player/rules#rule_key=typescript%3AS4325))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=rx-player&pullRequest=187&issues=AWL4yfupmtQEh1117EeV&open=AWL4yfupmtQEh1117EeV)

[ISSUE_KEY]: AWL4yfupmtQEh1117EeV